public-docs/environment-variables.md

```diff
@@ -0,0 +1,100 @@
+- `REACTION_LOG_LEVEL`: Optional. If you don't set this, the development default is "DEBUG" and the production default is "WARN".
+- `REACTION_SHOULD_INIT_REPLICA_SET`: Optional. If this is `true`, on startup the API will auto-initialize a MongoDB replica set if one isn't found. This is `true` by default when `NODE_ENV` is `development`; otherwise, it's `false`.
+- `REACTION_WORKERS_ENABLED`: Optional. Set to `false` to disable background job workers. Be careful because at least one instance must be working background jobs or features such as emailing and file uploads won't work.
+- `ROOT_URL`: Required. The root URL with protocol, such as `https://shop.mydomain.me`.
+- `STRIPE_API_KEY`: Required if you want Stripe payments to work. If you aren't using Stripe but you have the `payments-stripe` plugin registered, set this to any string.
+
+### Hydra
+
+> Other environment variables or configuration options may be available. Refer to the [Hydra documentation](https://www.ory.sh/docs/oryos.11/hydra/).
+
+- `DSN`: Required. The Postres database connection string.
+- `OAUTH2_EXPOSE_INTERNAL_ERRORS`: Optional. Useful for debugging. Refer to https://www.ory.sh/docs/next/hydra/debugging#first-aid
+- `OIDC_SUBJECT_IDENTIFIERS_ENABLED`: Required. Be careful not to change this in production. Refer to https://www.ory.sh/docs/oryos.9/hydra/advanced#openid-connect
+- `OIDC_SUBJECT_IDENTIFIERS_PAIRWISE_SALT`: Required. Be careful not to change this in production. Refer to https://www.ory.sh/docs/oryos.9/hydra/advanced#openid-connect
+- `SECRETS_SYSTEM`: Required. A unique secret key.
+- `URLS_CONSENT`: Required. Reaction Identity consent URL (the root URL plus `/consent`).
+- `URLS_ERROR`: Required. Reaction Identity errors URL (the root URL plus `/account/oauth-error`).
+- `URLS_LOGIN`: Required. Reaction Identity login URL (the root URL plus `/login`).
+- `URLS_LOGOUT`: Required. Reaction Identity logout URL (the root URL plus `/logout`).
+- `URLS_SELF_ISSUER`: Required. The public URL of Hydra.
+
+### Example Storefront
+
+- `CANONICAL_URL`: Required. The canonical root public URL for your storefront.
+- `ENABLE_SPA_ROUTING`: Optional. By default, the example storefront uses in-browser routing as you click links to move around the site. In other words, the pages are not requested from the NextJS server but rather generated locally. If you need all links to do a full page refresh, then set this to "false".
+- `EXTERNAL_GRAPHQL_URL`: Required. Ensure this is the public API URL.
+- `INTERNAL_GRAPHQL_URL`: Required. Ensure this is the internal API URL.
+- `NODE_ENV`: When running locally, you must be sure that `NODE_ENV` is NOT in your `.env` file. For deployed environments (including non-production), be sure to set `NODE_ENV` to `production` or NextJS will run in development mode, which makes pages take a long time to load.
+- `OAUTH2_ADMIN_PORT`: Required. The internal Hydra admin port, which is `4445` unless you have changed it.
```

Comment:
I think we might not be consistently using the words "Required" and "Optional" in this doc. To me, any value with a built-in default is by that definition optional for the user to override. Thus any `*_PORT` env var should be optional.

Reply:
@focusaurus This documentation is correct. We have no default for this variable so it's required. I'm simply trying to help them out by stating in the docs that most likely it is `4445` because that's the default on the Hydra side. Maybe there's a way to make that clearer.

That said, this one is technically only required for local dev because only the `bin/start` script uses it. I could update that script to use `4445` if it's missing, and update these docs to state that it's only required when you're running locally and you've changed it.

Reply:
Yeah I think we should have defaults. Currently a quick browser search of this PR indicates we might have like 40 "required" env vars and I'm not following what's going on. Basically I'm looking for clarity between the cases of we have a default that should work in local dev, you might want to change this in prod, you definitely want to change this in prod, and there's no default this must be manually set, etc.

Reply:
@focusaurus All ENV have defaults for local dev. I think I mentioned this somewhere in one of these articles. So the Required/Optional here is referring to deployments. That's generally the audience here. I will look for a way to make this clearer.

Reply:
Well, I guess it's not quite that simple. Some have defaults in code and others have defaults for local dev in that they are defined in `.env.example` with the assumption you're running all projects on default ports, and that is copied to `.env` and used. But I'm combining those two things when I say "defaults for local dev".